### PR TITLE
Payment Channel

### PR DESCRIPTION
**Pull request by rllola**
Payment channels fuctionalities for rust/wasm/js.

### CODE REVIEW

File: examples/wasm_browser/index.js [hunk 1]
```diff
@@ -1,10 +1,13 @@
...
+//import * as wasm from "@zondax/filecoin-signing-tools/js";
+import * as wasm from "../../signer-npm/pkg/browser/filecoin_signer_wasm.js";
+//import * as wasm from "../../signer-npm/pkg/browser/filecoin_signer_wasm_bg.js";
+//import TransportWebUSB from "@ledgerhq/hw-transport-webusb";
 
 function log(text) {
   document.getElementById("output").innerHTML += text + "\n";
 }
 
+/*
```
Comment (jleni) [inline examples/wasm_browser/index.js:10]: Should we split this into two examples?

File: examples/wasm_node/multisig.js [hunk 2]
```diff
@@ -0,0 +1,175 @@
+const filecoin_signer = require('@zondax/filecoin-signer');
+const bip39 = require('bip39');
+const bip32 = require('bip32');
+const axios = require('axios');
+
+const URL = "http://192.168.1.38:1234/rpc/v0"
+const TOKEN = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.2l5dql9PvaFPNY8_0UynbY0usJZ99dbIwpPUtaEDIYs"
```
Comment (jleni) [inline examples/wasm_node/multisig.js:7]: Let's avoid any tokens here as per audit feedback

File: examples/wasm_node/payment_channel.js [hunk 3]
```diff
@@ -0,0 +1,363 @@
+const filecoin_signer = require('@zondax/filecoin-signing-tools');
+const bip39 = require('bip39');
+const bip32 = require('bip32');
+const axios = require('axios');
+
+const URL = "http://192.168.1.38:1234/rpc/v0"
+const TOKEN = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.1Ys00_AiuYuXpNQ_eexBNKL1ZYq_6Xim9SpYwXs3GwY"
```
Comment (jleni) [inline examples/wasm_node/payment_channel.js:7]: Same here, let's remove tokens, let's use environment

File: examples/wasm_node/test/tests.js [hunk 4]
```diff
@@ -912,6 +913,277 @@ describeCall('SerializeParams', function () {
...
+    
+    console.log(create_pymtchan)
+    
+    let signature = filecoin_signer.transactionSignLotus(create_pymtchan, privateKey);
+
+    console.log(JSON.parse(signature));
+
+    assert.deepStrictEqual(expected.Message, JSON.parse(signature).Message);
+    
+    // TODO: verify signature
```
Comment (jleni) [inline examples/wasm_node/test/tests.js:1015]: Yep, I think we need to open an issue for this.

File: extras/src/multisig.rs [hunk 5]
```diff
@@ -0,0 +1,100 @@
+use clock::ChainEpoch;
+use forest_address::Address;
+use forest_encoding::tuple::*;
+use forest_vm::{MethodNum, Serialized, TokenAmount, METHOD_CONSTRUCTOR};
+use num_bigint::biguint_ser;
+use serde::{Deserialize, Serialize};
+
+/// Transaction ID type
+// TODO change to uvarint encoding
```
Comment (jleni) [inline extras/src/multisig.rs:9]: Is this old of should we create an issue for this?

File: signer-npm/src/lib.rs [hunk 6]
```diff
@@ -431,6 +429,131 @@ pub fn cancel_multisig(
...
+#[wasm_bindgen(js_name = updatePymtChan)]
+pub fn update_pymtchan(
+    pch_address: String,
+    from_address: String,
+    signed_voucher: String,
+    nonce: u32,
+) -> Result<JsValue, JsValue> {
+    set_panic_hook();
+
+    // TODO: verify if `pch_address` is an actor address. Not needed but good improvement.
```
Comment (jleni) [inline signer-npm/src/lib.rs:499]: We have a problem with the amount of time this takes..
Maybe it is a good idea to create an issue for this improvement

File: signer/src/api.rs [hunk 7]
```diff
@@ -259,32 +269,205 @@ pub struct ChangeNumApprovalsThresholdMultisigParams {
...
+        serde_bytes::Serialize::serialize(&v, serializer)
+    }
+}
+
+/// Signed voucher
+#[cfg_attr(feature = "with-arbitrary", derive(arbitrary::Arbitrary))]
+//#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
+#[derive(Debug, Clone, PartialEq, Deserialize_tuple, Serialize_tuple)]
+pub struct SignedVoucherAPI {
+    pub time_lock_min: i64, // not supported - must be 0
```
Comment (jleni) [inline signer/src/api.rs:375]: Let's create an issue for these things. Or at least let's put them in gdoc.

File: signer/src/lib.rs [hunk 8]
```diff
@@ -674,30 +673,254 @@ pub fn serialize_params(params: MessageParams) -> Result<CborBuffer, SignerError
...
+    let message_params_create_pymtchan = ExecParams {
+        code_cid: Cid::new_v1(Codec::Raw, Identity::digest(b"fil/1/paymentchannel")),
+        constructor_params: serialized_constructor_params,
+    };
+
+    let serialized_params =
+        forest_vm::Serialized::serialize::<ExecParams>(message_params_create_pymtchan)
+            .map_err(|err| SignerError::GenericString(err.to_string()))?;
+
+    // TODO:  don't hardcode gas limit and gas price; use a gas estimator!
```
Comment (jleni) [inline signer/src/lib.rs:709]: Yes, this can be part of the upgrade to space race testnet

File: signer/src/lib.rs [hunk 9]
```diff
@@ -1053,6 +1276,421 @@ mod tests {
...
+        let privkey = PrivateKey::try_from(from_key).unwrap();
+
+        let _pch_create_message_unsigned = serde_json::json!(PYMTCHAN_EXAMPLE_UNSIGNED_MSG);
+        let pch_create_message_api: UnsignedMessageAPI =
+            serde_json::from_str(PYMTCHAN_EXAMPLE_UNSIGNED_MSG)
+                .expect("Could not serialize unsigned message");
+        // TODO:  ^^^ this is an error, these lines are duplicated.  First one should have called create_pymtchan()
+
+        let signed_message_result = transaction_sign(&pch_create_message_api, &privkey).unwrap();
+        // TODO:  how do I check the signature of a transaction_sign() result
```
Comment (jleni) [inline signer/src/lib.rs:1452]: Let's chat about this one

File: signer/src/lib.rs [hunk 10]
```diff
@@ -1053,6 +1276,421 @@ mod tests {
...
+                0x0C, 0x83, 0xE5, 0xCE, 0x59, 0xD4, 0xF6, 0x1D, 0x01,
+            ],
+        };
+        let sv = SignedVoucherAPI::new(0, 1, 1, &sig);
+        let serialized_cbor = to_vec_packed(&sv).unwrap();
+        let mut serialized_cbor_hex_string = String::new();
+        serialized_cbor
+            .iter()
+            .for_each(|x| serialized_cbor_hex_string.push_str(&format!("{:02x}", x)));
+        //println!("serialized cbor = '{}'",serialized_cbor_hex_string);
```
Comment (jleni) [inline signer/src/lib.rs:1685]: we could move to a logger here as in other tests

File: signer/src/lib.rs [hunk 11]
```diff
@@ -1053,6 +1276,421 @@ mod tests {
...
+        assert_eq!(serialized_cbor_hex_string, test_vector_cbor_hex_string);
+    }
+
+    /*#[test]
+    fn test_sign_voucher() {
```
Comment (jleni) [inline signer/src/lib.rs:1690]: mm.. Maybe mark this test as skip, so we keep it mind?

**Review by jleni — CHANGES_REQUESTED** [review]

**Comment by rllola** [comment]
Outdated.

File: extras/src/multisig.rs (hunk 5 above)
Comment (rllola) [reply extras/src/multisig.rs:9]: This come from here but I don't know what it means really : https://github.com/Zondax/forest/blob/master/vm/actor/src/builtin/multisig/types.rs#L12-L15

Comment (rllola) [reply extras/src/multisig.rs:9]: https://github.com/Zondax/filecoin-signing-tools/issues/270

**Pull request closed without merging**